### PR TITLE
feat: manage vault in api

### PR DESCRIPTION
**Pull request by djaiss**

### CODE REVIEW

File: tests/Unit/Domains/Vault/ManageVault/Api/Controllers/VaultControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+        $form = [
+            'account_id' => $user->account_id,
+            'author_id' => $user->id,
+            'vault_id' => $vault->id,
+            'name' => 'this is a name',
+            'description' => 'this is a cool description',
+        ];
+
+        $response = $this->put('/api/vaults/'.$vault->id, $form);
```
Comment (asbiin) [inline tests/Unit/Domains/Vault/ManageVault/Api/Controllers/VaultControllerTest.php:136]: Good test: change the `Carbon` date before this to see if changes the `updated_at`?

File: routes/api.php [hunk 2]
```diff
@@ -10,4 +11,16 @@
...
+    // vaults
+    Route::prefix('vaults')->group(function () {
+        Route::get('', [VaultController::class, 'index'])->name('vaults.index')->middleware('abilities:read');
+        Route::post('', [VaultController::class, 'store'])->name('vaults.store')->middleware('abilities:read');
+
+        Route::prefix('{vault}')->group(function () {
+            Route::get('', [VaultController::class, 'show'])->name('vaults.show');
+            Route::put('', [VaultController::class, 'update'])->name('vaults.update');
+            Route::delete('', [VaultController::class, 'destroy'])->name('vaults.destroy');
+        });
```
Comment (asbiin) [inline routes/api.php]: I would really like us to use the `Route::apiResource` method for these kind of routes, and it is so easy to use.

**Review by asbiin — CHANGES_REQUESTED** [review]

File: tests/Unit/Domains/Vault/ManageVault/Api/Controllers/VaultControllerTest.php (hunk 1 above)
Comment (djaiss) [reply tests/Unit/Domains/Vault/ManageVault/Api/Controllers/VaultControllerTest.php:136]: done

File: routes/api.php (hunk 2 above)
Comment (djaiss) [reply routes/api.php]: done

File: phpstan.neon [hunk 3]
```diff
@@ -14,6 +14,7 @@ parameters:
         # psalm/plugin-laravel doesn't support \Illuminate\Database\Eloquent\Casts\Attribute yet
         - '#Access to an undefined property App\\Models\\Contact.*::\$name\.#'
         - '#Access to an undefined property App\\Http\\Resources\\UserResource::\$[a-zA-Z0-9_]+\.#'
+        - '#Access to an undefined property App\\Http\\Resources\\VaultResource::\$[a-zA-Z0-9_]+\.#'
```
Comment (asbiin) [inline phpstan.neon]: I think we should be able to group that like:
```suggestion
        - '#Access to an undefined property App\\Http\\Resources\\[^:]+::\$[a-zA-Z0-9_]+\.#'
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+use App\Domains\Vault\ManageVault\Services\UpdateVault;
+use App\Http\Controllers\ApiController;
+use App\Http\Resources\VaultResource;
+use App\Models\Vault;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+use Illuminate\Database\QueryException;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+
+class VaultController extends ApiController
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php:21]: We can add the group here
```suggestion
/**
 * @group Vault management
 * @subgroup Vaults
 */
class VaultController extends ApiController
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    /**
+     * List all vaults
+     *
+     * Get all the vaults in the account.
+     *
+     * @group Vault management
+     * @subgroup Vaults
+     * @queryParam limit int A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10. Example: 10
+     * @apiResourceModel App\Models\Vault
+     */
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]: Lets try with the attributes:
```suggestion
     */
    #[QueryParam('limit', 'int', description: 'A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10.', required: false, example: 10)]
    #[ResponseFromApiResource(VaultResource::class, Vault::class, collection: true)]
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+     * Create a vault
+     *
+     * Creates a vault object.
+     *
+     * @group Vault management
+     * @subgroup Vaults
+     * @bodyParam name string required The name of the vault. Max 255 characters.
+     * @bodyParam description string The description of the vault. Max 65535 characters.
+     * @apiResourceModel App\Models\Vault
+     */
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]:
```suggestion
     */
    #[BodyParam('name', description: 'The name of the vault. Max 255 characters.')]
    #[BodyParam('description', description: 'The description of the vault. Max 65535 characters.', required: false)]
    #[ResponseFromApiResource(VaultResource::class, Vault::class, status: 201)]
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    /**
+     * Retrieve a vault
+     *
+     * Get a specific vault object.
+     *
+     * @group Vault management
+     * @subgroup Vaults
+     * @apiResourceModel App\Models\Vault
+     */
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]:
```suggestion
     */
    #[ResponseFromApiResource(VaultResource::class, Vault::class)]
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+     * If the call succeeds, the response is the same as the one for the
+     * Retrieve a vault endpoint.
+     *
+     * @group Vault management
+     * @subgroup Vaults
+     * @urlParam id int required The vault's ID.
+     * @bodyParam name string required The name of the vault. Max 255 characters.
+     * @bodyParam description string The description of the vault. Max 65535 characters.
+     * @apiResourceModel App\Models\Vault
+     */
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]:
```suggestion
     */
    #[BodyParam('name', description: 'The name of the vault. Max 255 characters.')]
    #[BodyParam('description', description: 'The description of the vault. Max 65535 characters.', required: false)]
    #[ResponseFromApiResource(VaultResource::class, Vault::class)]
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+     * Delete a vault
+     *
+     * Destroys a vault object.
+     * Warning: everything in the vault will be immediately deleted.
+     *
+     * @group Vault management
+     * @subgroup Vaults
+     * @response status=200 {"deleted": "true", "id": 1}
+     * @urlParam id int required The vault's ID.
+     */
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]:
```suggestion
     */
    #[Response(["deleted" => true, "id" => 1])]
```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+use App\Domains\Vault\ManageVault\Services\DestroyVault;
+use App\Domains\Vault\ManageVault\Services\UpdateVault;
+use App\Http\Controllers\ApiController;
+use App\Http\Resources\VaultResource;
+use App\Models\Vault;
+use Illuminate\Database\Eloquent\ModelNotFoundException;
+use Illuminate\Database\QueryException;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]: Below changes requires these uses:
```suggestion
use Knuckles\Scribe\Attributes\BodyParam;
use Knuckles\Scribe\Attributes\QueryParam;
use Knuckles\Scribe\Attributes\ResponseFromApiResource;
use Knuckles\Scribe\Attributes\Response;

```

File: app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+class VaultController extends ApiController
+{
+    public function __construct()
+    {
+        $this->middleware('abilities:read')->only(['index', 'show']);
+        $this->middleware('abilities:create')->only(['store']);
+        $this->middleware('abilities:update')->only(['update']);
+        $this->middleware('abilities:destroy')->only(['delete']);
```
Comment (asbiin) [inline app/Domains/Vault/ManageVault/Api/Controllers/VaultController.php]:
```suggestion
        $this->middleware('abilities:write')->only(['store', 'update', 'delete']);
```

**Review by asbiin — CHANGES_REQUESTED** [review]
Small changes according to recent commits

**Review by asbiin — APPROVED** [review]

**Comment by djaiss** [comment]
Incredible.

**Pull request merged**